### PR TITLE
Add --ios-gl option to compile against iOS OpenGLES implementation

### PR DESCRIPTION
**Pull request by tmm1**
This is a pretty ugly patch, but it makes all the OpenGL code compile against the OpenGLES headers that ship with iOS.

Ideally we would also teach `wscript` a new `--ios` option to actually build this correctly, instead of having to drag/drop files into an Xcode project. To compile this code with the xcode toolchain on the command line, something like this should work:

```
export PATH="/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/:$PATH"
export SDKPATH="$(xcodebuild -sdk iphoneos -version Path)"
export CFLAGS="-isysroot $SDKPATH -arch arm64 -mios-version-min=9.0.0 -fembed-bitcode"
export LDFLAGS="-isysroot $SDKPATH -arch arm64 -Wl,-ios_version_min,9.0.0 -lbz2"

./waf configure --disable-cplayer --disable-lcms2 --disable-lua --enable-libmpv-static --enable-ios-gl --enable-gl
./waf build
```

Ofcourse, in reality you would probably also want to build a armv7 binary, and an x86_64 binray for the ios-simulator, and then you'd `lipo` them all together. So it gets complicated pretty quickly. I'm not sure if all this logic would be best added into `wscript`, or an external script which invoke waf multiple times.

### CODE REVIEW

File: video/out/opengl/video.c [hunk 1]
```diff
@@ -679,7 +685,7 @@ static bool gl_video_get_lut3d(struct gl_video *p, enum mp_csp_prim prim,
 
     gl->BindTexture(GL_TEXTURE_3D, p->lut_3d_texture);
     gl->PixelStorei(GL_UNPACK_ALIGNMENT, 1);
-    gl->TexImage3D(GL_TEXTURE_3D, 0, GL_RGB16, lut3d->size[0], lut3d->size[1],
+    gl->TexImage3D(GL_TEXTURE_3D, 0, GL_RGB16F, lut3d->size[0], lut3d->size[1],
```
Comment (tmm1) [inline video/out/opengl/video.c]: This change should clearly be `ifdef`d as well.

**Comment by tmm1** [comment]
Many of the `ifdef` in this patch guard against symbols that are missing in the iOS headers..

```
$ cd /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/
$ grep -r GL_RGBA12 .
$
```

Others are unavailable, but have `_I` or `_EXT` variants that we might be able to `#define` and use instead.

```
$ grep -r GL_LUMINANCE8 .
./OpenGLES.framework/Headers/ES1/glext.h:#define GL_LUMINANCE8_EXT                                       0x8040
./OpenGLES.framework/Headers/ES1/glext.h:#define GL_LUMINANCE8_ALPHA8_EXT                                0x8045
./OpenGLES.framework/Headers/ES2/glext.h:#define GL_LUMINANCE8_EXT                                       0x8040
./OpenGLES.framework/Headers/ES2/glext.h:#define GL_LUMINANCE8_ALPHA8_EXT                                0x8045
```

```
$ grep -r GL_R16 .
./OpenGLES.framework/Headers/ES2/glext.h:#define GL_R16F_EXT                                             0x822D
./OpenGLES.framework/Headers/ES3/gl.h:#define GL_R16F                                          0x822D
./OpenGLES.framework/Headers/ES3/gl.h:#define GL_R16I                                          0x8233
./OpenGLES.framework/Headers/ES3/gl.h:#define GL_R16UI                                         0x8234
./OpenGLES.framework/Headers/ES3/glext.h:#define GL_R16F_EXT                                             0x822D
```

**Comment by jeeb** [comment]
In case of Android quite a bit of missing defines were added to a compatibility header under `#if HAVE_ANDROID_GL`  @ https://github.com/mpv-player/mpv/pull/2805/commits/7956d8c91a1d6f53ef79640976a630df2a73aaed . I would guess that is a valid way of dealing with iOS oddities as well?

**Comment by tmm1** [comment]
> In case of Android quite a bit of missing defines were added to a compatibility header under #if HAVE_ANDROID_GL @ 7956d8c . I would guess that is a valid way of dealing with iOS oddities as well?

Thanks for the pointer. That does seem like a much better approach, and would clean up all the ugly ifdef in this diff.

Are the values for the missing constants generally the same across platforms? I'm not sure how I would find them for iOS.

**Comment by jeeb** [comment]
IIRC the headers are standardized so I would say the values should be the same between implementations.

**Comment by tmm1** [comment]
I redid the diff in here based on @jeeb's advice. All the `ifdef` are gone and the new patch is significantly smaller and cleaner!

**Comment by tmm1** [comment]
@pigoz Yea, the only way to use videotoolbox is with `--hwdec=videotoolbox-copy`. I think there's an iOS equivalent of IOSurface that can be used to do zero-copy rendering, but I haven't looked into it.

**Comment by wm4** [comment]
> @pigoz Yea, the only way to use videotoolbox is with --hwdec=videotoolbox-copy. I think there's an iOS equivalent of IOSurface that can be used to do zero-copy rendering, but I haven't looked into it.

Well, that's sad. Other than that, this looks reasonable, or at least not worse than the Android case. The change to the HAVE_C11_TLS `#if` check makes no sense to me, though.

**Comment by tmm1** [comment]
> The change to the HAVE_C11_TLS #if check makes no sense to me, though.

The iOS compiler does not support thread locals.

**Comment by wm4** [comment]
But then the configure check should catch this somehow.

**Comment by tmm1** [comment]
> But then the configure check should catch this somehow.

I went to check why configure wasn't catching it, but it looks like the newly released Xcode 8 now implements thread locals. I have updated the PR accordingly.

**Comment by wm4** [comment]
Looks great now - merged.

**Comment by XiaoChenYung** [comment]
image
I create libmpv.a in xcode with ios，but when compile i get many errors

**Comment by antonz0319** [comment]
@XiaoChenYung , Did you solve your problem?

**Comment by MenThu** [comment]
Compile the iOS static library according to the script you provided and then I encountered this error:
`Unable to find either POSIX or MinGW-w64 environment, or compiler does not work.`
Full log:[waf_full_log.log](https://github.com/mpv-player/mpv/files/7755618/waf_full_log.log)
Waf version number:  `waf 2.0.20 (668769470956da8c5b60817cb8884cd7d0f87cd4)`

Regards

**Pull request closed without merging**